misk/src/main/kotlin/misk/web/actions/ReadinessCheckAction.kt

```diff
@@ -1,48 +1,42 @@
 package misk.web.actions
 
-import com.google.common.util.concurrent.ServiceManager
-import misk.healthchecks.HealthCheck
 import misk.security.authz.Unauthenticated
 import misk.web.AvailableWhenDegraded
 import misk.web.Get
 import misk.web.Response
 import misk.web.ResponseContentType
+import misk.web.WebConfig
 import misk.web.mediatype.MediaTypes
 import wisp.logging.getLogger
+import java.time.Clock
 import javax.inject.Inject
-import javax.inject.Provider
 import javax.inject.Singleton
 
 private val logger = getLogger<ReadinessCheckAction>()
 
 @Singleton
 class ReadinessCheckAction @Inject internal constructor(
-  private val serviceManagerProvider: Provider<ServiceManager>,
-  @JvmSuppressWildcards private val healthChecks: List<HealthCheck>
+  private val readinessService: ReadinessCheckService,
+  private val config: WebConfig,
+  private val clock: Clock,
 ) : WebAction {
 
   @Get("/_readiness")
   @ResponseContentType(MediaTypes.APPLICATION_JSON)
   @Unauthenticated
   @AvailableWhenDegraded
   fun readinessCheck(): Response<String> {
-    val servicesNotRunning = serviceManagerProvider.get().servicesByState().values().asList()
-      .filterNot { it.isRunning }
 
-    for (service in servicesNotRunning) {
-      logger.info("Service not running: $service")
-    }
+    val (lastUpdate, statuses) = readinessService.status
+      // Null until first run has finished, which means not yet ready
+      ?: return Response("", statusCode = 503)
 
-    if (!servicesNotRunning.isEmpty()) {
-      // Don't do healthchecks if services haven't all started. The app isn't in a good state yet,
-      // and a health check could end up triggering random errors that we don't want to flood the
-      // logs with.
+    if (clock.instant().isAfter(lastUpdate.plusMillis(config.readiness_max_age_ms.toLong()))) {
```

Comment:
should we push all this logic to the `readinessService`?

Reply:
I left this in the action because its a check against slowness or failures from the service. To put this in the service it would need to be on the getter... but then what would it do? Throw an error? Return null (which would be indistinguishable from "has not yet run")?

I think it makes more sense here.